src/websock/native/integrate_test.go

```diff
@@ -0,0 +1,508 @@
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(clientTimeout * time.Millisecond)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+		It("limits a call to the server by the server timeout when the clients' timeout is longer", func() {
+			deadline := make(chan time.Time)
+
+			peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+				// never return anything
+
+				dead, _ := ctx.Deadline()
+				deadline <- dead
+			})
+
+			// client timeouts are in milliseconds
+			clientTimeout := time.Duration(10000)
+			serverTimeout := time.Duration(2000) * time.Millisecond
+
+			subject = native.NewHandler(peer, message.JSONEncoding, native.MaxCallTimeout(serverTimeout))
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(serverTimeout)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+		It("limits a call to the server by the clients' timeout when the servers' timeout is not set", func() {
+			deadline := make(chan time.Time)
+
+			peer.Listen(ns, func(ctx context.Context, req rinq.Request, res rinq.Response) {
+				// never return anything
+
+				dead, _ := ctx.Deadline()
+				deadline <- dead
+			})
+
+			// client timeouts are in milliseconds
+			clientTimeout := time.Duration(10000)
+
+			subject = native.NewHandler(peer, message.JSONEncoding)
+
+			websocket.queueMsg(createSession, session, nil, nil)
+			websocket.queueMsg(callSync, session, []interface{}{
+				seq, ns, cmd, clientTimeout,
+			}, "ping")
+
+			expectedTime := time.Now().Add(clientTimeout * time.Millisecond)
+
+			close(start)
+
+			Expect(<-deadline).To(BeTemporally("~", expectedTime, time.Second/2))
+		})
+
+	})
+})
+
+func msg(msgType uint16, session uint16, headers interface{}, payload interface{}) []byte {
+
+	buff := new(bytes.Buffer)
+
+	b := make([]byte, 4)
+	binary.BigEndian.PutUint16(b[:2], msgType)
+	binary.BigEndian.PutUint16(b[2:], session)
+
+	buff.Write(b)
+
+	if headers != nil {
+		if err := message.JSONEncoding.EncodeHeader(buff, headers); err != nil {
+			panic(err)
+		}
+
+		if payload != noBody {
+			p := rinq.NewPayload(payload)
+			if err := message.JSONEncoding.EncodePayload(buff, p); err != nil {
+				panic(err)
+			}
+		}
+	}
+
+	return buff.Bytes()
+}
+
+type mockWebsock struct {
+	start <-chan struct{}
+	dead  <-chan struct{}
+
+	rOut []io.Reader
+	wIn  chan []byte
+}
+
+func (m *mockWebsock) queueMsg(msgType uint16, session uint16, headers interface{}, payload interface{}) {
+	out := msg(msgType, session, headers, payload)
+	(*m).rOut = append((*m).rOut, bytes.NewBuffer(out))
```

Comment:
Can `(*m).rOut` not just be written `m.rOut`, or am I missing something subtle here?

Reply:
was initially a single `[]io.Reader`, is just "legacy cruft". I'll flick it over.